### PR TITLE
Fix: add monotonic estimation for DateTime64 in toDayOfWeek

### LINKED ISSUE #71810: toDayOfWeek on WHERE condition with primary key breaks on updating clickhouse (Bad get)
**Describe the unexpected behaviour**
using `toDayOfWeek` as part of a WHERE condition in a select query fails after updating clickhouse to 24.8+.

reproduce here:
https://fiddle.clickhouse.com/8a03ec44-0ff7-4e45-8e85-8f3d9159e72c
fails on 24.8, succeeds on 24.7.

The error does not reproduce when the table is not ordered by the timestamp (https://fiddle.clickhouse.com/ee226b86-231d-4f6d-9815-14b0eab341d3), or when using a Memory table engine. Using this function in the SELECT part works. Also, similar functions like `toDayOfMonth` seem to work as usual.

Thanks!

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix exception for toDayOfWeek on WHERE condition with primary key of DateTime64 type.

closes #71810 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit f7c46114f4ca74d7afc1d7e335a469e8432f459e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71849/f7c46114f4ca74d7afc1d7e335a469e8432f459e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d38bd0df0aa277f4ec33b58f6832e45de596525c/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Functions/IFunctionCustomWeek.h [hunk 1]
```diff
@@ -56,6 +56,21 @@ class IFunctionCustomWeek : public IFunction
                 : is_not_monotonic;
         }
 
+        if (checkAndGetDataType<DataTypeDateTime64>(&type))
+        {
+
+            const auto & left_date_time = left.safeGet<DateTime64>();
+            TransformDateTime64<Transform> transformer_left(left_date_time.getScale());
```
Comment (Michicosun) [inline src/Functions/IFunctionCustomWeek.h:63]: As far as I understand, you need to use Transform::FactorTransform in this place, or am I wrong?

**Review by Michicosun — CHANGES_REQUESTED** [review]

File: src/Functions/IFunctionCustomWeek.h (hunk 1 above)
Comment (yakov-olkhovskiy) [reply src/Functions/IFunctionCustomWeek.h]: oh, yes, exactly... fixed

**Review by Michicosun — APPROVED** [review]

**Comment by yakov-olkhovskiy** [comment]
inconsistent select in 03210_variant_with_aggregate_function_type reported:
https://github.com/ClickHouse/ClickHouse/issues/72008

**Pull request merged**